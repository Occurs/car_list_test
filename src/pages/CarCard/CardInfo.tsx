import React, { useState, useEffect } from "react";
import { makeStyles } from "@material-ui/core/styles";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";
import Grid from "@material-ui/core/Grid";
import Box from "@material-ui/core/Box";
import { ICar } from "types/types";
import { capitalize } from "utils/capitalize";
import { colors as colorPallette } from "styles/variables";

type ICardInfo = {
  car: ICar;
};

const useStyles = makeStyles({
  image: {
    width: "100%",
    height: "inherit",
    maxHeight: "300px",
    objectFit: "contain",
  },
  button: {
    marginLeft: "auto",
    marginRight: "0px",
    display: "block",
    color: "#FFFFFF !important",
    width: "128px",
  },
});

const CarCardPage = ({ car }: ICardInfo) => {
  const classes = useStyles();
  const [favoriteCars, setFavoriteCars] = useState<Array<ICar>>([]);

  useEffect(() => {
    const favoriteCarsList = localStorage.getItem("favoriteCarsList");
    if (favoriteCarsList !== null) {
      const parsedList = JSON.parse(favoriteCarsList);
      setFavoriteCars(parsedList);
    }
  }, [car.stockNumber]);

  function addToLocalStorage() {
    const updatedFavoriteCars = [...favoriteCars, car];
    setFavoriteCars(updatedFavoriteCars);
    localStorage.setItem(
      "favoriteCarsList",
      JSON.stringify(updatedFavoriteCars)
    );
  }

  function removeFromLocalStorage() {
    const favCars = [...favoriteCars];
    const index = favCars.findIndex(
      (favoriteCar) => favoriteCar.stockNumber === car.stockNumber
    );
    if (index > -1) {
      favCars.splice(index, 1);
      setFavoriteCars(favCars);
      localStorage.setItem("favoriteCarsList", JSON.stringify(favCars));
    }
  }

  function checkInFavorites() {
    return (
      favoriteCars.filter(
        (favoriteCar) => favoriteCar.stockNumber === car.stockNumber
      ).length > 0
    );
  }

  return (
    <>
      <img
        alt={`${car.stockNumber}`}
        src={car.pictureUrl}
        className={classes.image}
      />
      <Grid
        container
        direction="row"
        justify="space-between"
        alignItems="stretch"
      >
        <Grid item xs={8}>
          <Box
            component="div"
            marginLeft="12px"
            marginTop="24px"
            height="100%"
            display="flex"
            alignItems="flex-start"
            flexDirection="column"
            justifyContent="space-between"
          >
            <Typography>
              <Box
                component="span"
                fontSize="32px"
                fontWeight="bold"
                margin="12px 0"
              >
                {`${car.manufacturerName} ${car.modelName}`}
              </Box>
            </Typography>
            <Typography>
              <Box component="span" fontSize="18px" margin="12px 0">
                {`Stock # ${car.stockNumber} - ${
                  car.mileage.number
                } ${car.mileage.unit.toUpperCase()} - ${
                  car.fuelType
                } - ${capitalize(car.color)}`}
              </Box>
            </Typography>
            <Typography>
              <Box component="span" fontSize="14px">
                This car is currently available and can be delivered as soon as
                tomorrow morning. Please be aware that delivery times shown in
                this page are not definitive and may change due to bad weather
                conditions.
              </Box>
            </Typography>
          </Box>
        </Grid>
        <Grid item xs={4}>
          <Box
            component="div"
            border={`2px solid ${colorPallette.secondary}`}
            padding="24px"
            width="100%"
            marginLeft="24px"
            marginTop="24px"
          >
            <Typography>
              <Box component="span" fontSize="14px">
                If you like this car, click the button and save it in your
                collection of favourite items.
              </Box>
            </Typography>
            {checkInFavorites() ? (
              <Button
                variant="contained"
                color="primary"
                onClick={removeFromLocalStorage}
                className={classes.button}
              >
                Remove
              </Button>
            ) : (
              <Button
                variant="contained"
                color="primary"
                onClick={addToLocalStorage}
                className={classes.button}
              >
                Save
              </Button>
            )}
          </Box>
        </Grid>
      </Grid>
    </>
  );
};

export default CarCardPage;
