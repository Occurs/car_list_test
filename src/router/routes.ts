const routes = {
  carsList: '/',
  carCard: '/car/:stockNumber',
  favorites: '/favorites'
};

export default routes;