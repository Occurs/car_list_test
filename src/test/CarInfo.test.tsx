import React from "react";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import CardInfo from "pages/CarCard/CardInfo";
import routes from "router/routes";

const props = {
  car: {
    stockNumber: 41400,
    manufacturerName: "Fiat",
    modelName: "Marea",
    mileage: {
      number: 100141,
      unit: "km",
    },
    fuelType: "Diesel",
    color: "white",
    pictureUrl: "",
  },
};

const renderComponent = () =>
  render(
    <MemoryRouter initialEntries={[routes.carsList]}>
      <CardInfo {...props} />
    </MemoryRouter>
  );

describe("CarCard Component", () => {
  test("snapshot", () => {
    const component = renderer.create(
      <MemoryRouter initialEntries={[routes.carsList]}>
        <CardInfo {...props} />
      </MemoryRouter>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("render", async () => {
    const { getByText } = renderComponent();
    getByText("Fiat Marea");
    getByText("Stock # 41400 - 100141 KM - Diesel - White");
  });
});
