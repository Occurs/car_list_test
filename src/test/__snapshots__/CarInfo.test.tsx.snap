// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarCard Component snapshot 1`] = `
Array [
  <img
    alt="41400"
    className="makeStyles-image-1"
    src=""
  />,
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-space-between"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
    >
      <div
        className="MuiBox-root MuiBox-root-3"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          <span
            className="MuiBox-root MuiBox-root-4"
          >
            Fiat Marea
          </span>
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          <span
            className="MuiBox-root MuiBox-root-5"
          >
            Stock # 41400 - 100141 KM - Diesel - White
          </span>
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          <span
            className="MuiBox-root MuiBox-root-6"
          >
            This car is currently available and can be delivered as soon as tomorrow morning. Please be aware that delivery times shown in this page are not definitive and may change due to bad weather conditions.
          </span>
        </p>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
    >
      <div
        className="MuiBox-root MuiBox-root-7"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          <span
            className="MuiBox-root MuiBox-root-8"
          >
            If you like this car, click the button and save it in your collection of favourite items.
          </span>
        </p>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-2 MuiButton-containedPrimary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Save
          </span>
        </button>
      </div>
    </div>
  </div>,
]
`;
